src/swish/osi.c

```diff
@@ -466,9 +466,19 @@ static ptr write_stream_port(uptr port, ptr buffer, size_t start_index, uint32_t
 }
 
 static void close_stream_cb(uv_handle_t* handle) {
+  // This code does not check and unlock the write_callback and
+  // write_buffer because LibUV explicitly cancels write operations
+  // with UV_ECANCELED before processing this callback.
   stream_port_t* p = container_of(handle, stream_port_t, h.stream);
   ptr callback = p->close_callback;
+  ptr read_callback = p->read_callback;
+  ptr read_buffer = p->read_buffer;
   free(p);
+  if (read_callback) {
+    Sunlock_object(read_buffer);
+    Sunlock_object(read_callback);
+    osi_add_callback1(read_callback, Sfixnum(UV_EOF));
```

Comment:
Should we use UV_ECANCELED for consistency?

Reply:
I think that might create unexpected behavior from the Scheme port perspective. Where you might have received either a values or an eof-object, a cancel will cause an exception. At least in the cases I've been working through, my primary process has explicitly closed the port, and I expect an eof-object, not an exception.

Reply:
That seems reasonable.